File: abc286/e.py
n = int(input())
aa = list(map(int, input().split()))
ss = [
    input()
    for _ in range(n)
]

inf = 1 << 60
d = [
    [(inf, 0) for _ in range(n)]
    for _ in range(n)
]
for i in range(n):
    for j in range(n):
        if ss[i][j] == 'Y':
            d[i][j] = (1, -aa[j])
        

for k in range(n):
    for i in range(n):
        for j in range(n):
            d1 = d[i][j]
            d2 = (d[i][k][0] + d[k][j][0], d[i][k][1] + d[k][j][1])
            if d1 > d2:
                d[i][j] = d2

q = int(input())
for i in range(q):
    u, v = map(int, input().split())
    u -= 1
    v -= 1
    if d[u][v][0] == inf:
        print('Impossible')
    else:
        print(d[u][v][0], -d[u][v][1] + aa[u])

File: abc286/e.cpp
#include <bits/stdc++.h>
using namespace std;

long long d[301][301];
long long v[301][301];

const int INF = INT_MAX;

int main(void) {
    int i, j, k, n;
    cin >> n;
    
    vector<int> aa(n);
    vector<string> ss(n);

    for(i = 0; i < n; i++) {
        cin >> aa[i];
    }
    for(i = 0; i < n; i++) {
        cin >> ss[i];
    }

    for(i = 0; i < n; i++) {
        for(j = 0; j < n; j++) {
            if(ss[i][j] == 'Y') {
                d[i][j] = 1;
                v[i][j] = aa[j];
            } else {
                d[i][j] = INT_MAX;
            }
        }
    }
    
    for(k = 0; k < n; k++) {
        for(i = 0; i < n; i++) {
            for(j = 0; j < n; j++) {
                if(d[i][j] > d[i][k] + d[k][j]) {
                    d[i][j] = d[i][k] + d[k][j];
                    v[i][j] = v[i][k] + v[k][j];
                } else if(d[i][j] == d[i][k] + d[k][j]) {
                    v[i][j] = max(v[i][j], v[i][k] + v[k][j]);
                }
            }
        }
    }

    int q;
    cin >> q;
    for(i = 0; i < q; i++) {
        int u1, u2;
        cin >> u1 >> u2;
        u1--; u2--;

        if(d[u1][u2] == INF) {
            cout << "Impossible" << endl;
        } else {
            cout << d[u1][u2] << " " << (v[u1][u2] + aa[u1]) << endl;
        }
    }

    return 0;
}